File: 714-best-time-to-buy-and-sell-stock-with-transaction-fee/best-time-to-buy-and-sell-stock-with-transaction-fee.java
class Solution {
    public int maxProfit(int[] prices, int fee) {
        int n = prices.length; 
        int free = 0;
        int hold = -prices[0];

        for(int i = 1; i < n; i++) {
            int prevFree = free;
            free = Math.max(free, hold + prices[i] - fee);
            hold = Math.max(hold, prevFree - prices[i]);
        }
        return free;
    }
}

File: 714-best-time-to-buy-and-sell-stock-with-transaction-fee/best-time-to-buy-and-sell-stock-with-transaction-fee.py
class Solution:
    def maxProfit(self, prices: List[int], fee: int) -> int:
        free = [0] * len(prices)
        hold = [0] * len(prices)
        hold[0] = -prices[0]
        for i in range(1, len(prices)):
            free[i] = max(hold[i - 1] + prices[i] - fee, free[i - 1])
            hold[i] = max(hold[i - 1], free[i - 1] - prices[i])

        return free[-1]